picasso-compose/src/androidTest/java/com/squareup/picasso3/compose/PicassoPainterTest.kt

```diff
@@ -0,0 +1,197 @@
+import org.junit.runner.RunWith
+import kotlinx.coroutines.Dispatchers
+
+@RunWith(AndroidJUnit4::class)
+class PicassoPainterTest {
+
+  @get:Rule
+  val rule = createComposeRule()
+
+  @Test
+  fun firstFrameConsumesStateFromLayout() {
+    lateinit var lastRequest: Request
+    val context = InstrumentationRegistry.getInstrumentation().targetContext
+
+    val picasso = Picasso.Builder(context)
+      .callFactory { throw RuntimeException() }
+      .dispatchers(Dispatchers.Unconfined, Dispatchers.Unconfined)
+      .addRequestHandler(object : RequestHandler() {
+        override fun canHandleRequest(data: Request): Boolean = true
+        override fun load(picasso: Picasso, request: Request, callback: Callback) {
+          lastRequest = request
+          callback.onSuccess(Result.Bitmap(Bitmap.createBitmap(1, 1, ARGB_8888), LoadedFrom.MEMORY))
+        }
+      })
+      .build()
+    var size: IntSize by mutableStateOf(IntSize.Zero)
+    var drawn = false
+
+    rule.setContent {
+      CompositionLocalProvider(LocalDensity provides Density(1f)) {
+        val painter = picasso.rememberPainter {
+          it.load("http://example.com/")
+            // Headers are not part of a cache key, using a stable key to break cache
+            .stableKey("http://example.com/$size")
+            .addHeader("width", size.width.toString())
+            .addHeader("height", size.height.toString())
+        }
+        Canvas(
+          Modifier
+            .requiredSize(9.dp)
+            .onSizeChanged { size = it }
+        ) {
+          val canvasSize = this.size
+
+          with(painter) {
+            draw(canvasSize)
+          }
+          drawn = true
+        }
+      }
+    }
+
+    rule.waitUntil { drawn }
+
+    // Draw triggers request was made with size.
+    assertThat(lastRequest.headers?.toMultimap()).containsAtLeastEntriesIn(
+      mapOf(
+        "width" to listOf("9"),
+        "height" to listOf("9")
+      )
+    )
+  }
+
+  @Test
+  fun redrawDoesNotReexecuteUnchangedRequest() {
+    var requestCount = 0
+    val context = InstrumentationRegistry.getInstrumentation().targetContext
+    val picasso = Picasso.Builder(context)
+      .callFactory { throw RuntimeException() }
+      .dispatchers(Dispatchers.Unconfined, Dispatchers.Unconfined)
+      .addRequestHandler(object : RequestHandler() {
+        override fun canHandleRequest(data: Request): Boolean = true
+        override fun load(picasso: Picasso, request: Request, callback: Callback) {
+          requestCount++
+          callback.onSuccess(Result.Bitmap(Bitmap.createBitmap(1, 1, ARGB_8888), LoadedFrom.MEMORY))
+        }
+      })
+      .build()
+
+    var drawInvalidator by mutableStateOf(0)
+    var drawCount = 0
+    rule.setContent {
+      CompositionLocalProvider(LocalDensity provides Density(1f)) {
+        val painter = picasso.rememberPainter {
+          it.load("http://example.com/")
+        }
+        Canvas(Modifier.fillMaxSize()) {
```

Comment:
would it be helpful to use a named argument for the lambda here?

Reply:
I don't think so, its the only other argument and generally accepted that Canvas' lambda is DrawScope